### PR TITLE
Initial commit of KafkaToBigQuery template.

### PR DESCRIPTION
**Pull request by chrlarsen**
@sabhyankar initial commit for Kafka to BigQuery

### CODE REVIEW

File: v2/checkstyle/checkstyle.xml [hunk 1]
```diff
@@ -0,0 +1,319 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/checkstyle/checkstyle.xml]: Removed.

File: v2/checkstyle/suppressions.xml [hunk 2]
```diff
@@ -0,0 +1,36 @@
+<?xml version="1.0"?>
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/checkstyle/suppressions.xml]: Removed.

File: v2/kafka-to-bigquery/pom.xml [hunk 3]
```diff
@@ -0,0 +1,100 @@
...
+    <properties>
+        <autovalue.annotations.version>1.6.5</autovalue.annotations.version>
+        <autovalue.version>1.6.5</autovalue.version>
+        <bigquery.version>v2-rev438-1.25.0</bigquery.version>
+        <google-storage-api.version>v1-rev158-1.25.0</google-storage-api.version>
+        <kafka-clients.version>2.3.0</kafka-clients.version>
+    </properties>
+
+    <dependencies>
+        <dependency>
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/pom.xml]: Verified and removed.

File: v2/kafka-to-bigquery/pom.xml [hunk 4]
```diff
@@ -0,0 +1,100 @@
...
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>kafka-to-bigquery</artifactId>
+    <modelVersion>4.0.0</modelVersion>
+
+    <properties>
+        <autovalue.annotations.version>1.6.5</autovalue.annotations.version>
+        <autovalue.version>1.6.5</autovalue.version>
+        <bigquery.version>v2-rev438-1.25.0</bigquery.version>
+        <google-storage-api.version>v1-rev158-1.25.0</google-storage-api.version>
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/pom.xml]: It is not, removed.

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/package-info.java [hunk 5]
```diff
@@ -0,0 +1,20 @@
...
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Main pipelines contained in this repository.
+ */
+package com.google.cloud.teleport.v2.templates;
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/package-info.java]: Done, added.

File: v2/kafka-to-bigquery/pom.xml [hunk 6]
```diff
@@ -0,0 +1,100 @@
...
+    <parent>
+        <artifactId>dynamic-templates</artifactId>
+        <groupId>com.google.cloud.teleport.v2</groupId>
+        <version>1.0-SNAPSHOT</version>
+    </parent>
+    <artifactId>kafka-to-bigquery</artifactId>
+    <modelVersion>4.0.0</modelVersion>
+
+    <properties>
+        <autovalue.annotations.version>1.6.5</autovalue.annotations.version>
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/pom.xml]: Yes, will move to parent pom.xml.

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/values/package-info.java [hunk 7]
```diff
@@ -0,0 +1,20 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/values/package-info.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/values/FailsafeElement.java [hunk 8]
```diff
@@ -0,0 +1,112 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/values/FailsafeElement.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/SchemaUtils.java [hunk 9]
```diff
@@ -0,0 +1,68 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/SchemaUtils.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/JavascriptTextTransformer.java [hunk 10]
```diff
@@ -0,0 +1,356 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/JavascriptTextTransformer.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/ErrorConverters.java [hunk 11]
```diff
@@ -0,0 +1,201 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/ErrorConverters.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/coders/package-info.java [hunk 12]
```diff
@@ -0,0 +1,20 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/coders/package-info.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/coders/FailsafeElementCoder.java [hunk 13]
```diff
@@ -0,0 +1,105 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/coders/FailsafeElementCoder.java]: Done. "Commons" module has been added and relevant packages have been added.

File: v2/pom.xml [hunk 14]
```diff
@@ -0,0 +1,520 @@
+<!--~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/pom.xml]: Done, removed.

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/BigQueryConverters.java [hunk 15]
```diff
@@ -0,0 +1,118 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/BigQueryConverters.java]: Done

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/package-info.java [hunk 16]
```diff
@@ -0,0 +1,20 @@
+/*
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/common/package-info.java]: Done

**Comment by sabhyankar** [comment]
@chrlarsen - Can you rebase and resolve the merge conflicts?

**Comment by chrlarsen** [comment]
@sabhyankar rebase is complete

File: v2/common/pom.xml [hunk 17]
```diff
@@ -0,0 +1,58 @@
...
+            <version>${avro.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.cloud.teleport.v2</groupId>
+            <artifactId>commons</artifactId>
+            <version>1.0-SNAPSHOT</version>
+            <scope>compile</scope>
+        </dependency>
+    </dependencies>
+</project>
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/common/pom.xml]: Done

File: v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryConverters.java [hunk 18]
```diff
@@ -0,0 +1,118 @@
...
+        new ByteArrayInputStream(json.getBytes(StandardCharsets.UTF_8))) {
+      row = TableRowJsonCoder.of().decode(inputStream, Context.OUTER);
+
+    } catch (IOException e) {
+      throw new RuntimeException("Failed to serialize json to table row: " + json, e);
+    }
+
+    return row;
+  }
+}
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/BigQueryConverters.java]: Done

File: v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/JavascriptTextTransformer.java [hunk 19]
```diff
@@ -0,0 +1,356 @@
...
+      javascriptRuntime =
+          JavascriptRuntime.newBuilder()
+              .setFunctionName(functionName)
+              .setFileSystemPath(fileSystemPath)
+              .build();
+    }
+
+    return javascriptRuntime;
+  }
+}
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/common/src/main/java/com/google/cloud/teleport/v2/transforms/JavascriptTextTransformer.java]: Done

File: v2/common/src/main/java/com/google/cloud/teleport/v2/values/FailsafeElement.java [hunk 20]
```diff
@@ -0,0 +1,112 @@
...
+  @Override
+  public String toString() {
+    return MoreObjects.toStringHelper(this)
+        .add("originalPayload", originalPayload)
+        .add("payload", payload)
+        .add("errorMessage", errorMessage)
+        .add("stacktrace", stacktrace)
+        .toString();
+  }
+}
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/common/src/main/java/com/google/cloud/teleport/v2/values/FailsafeElement.java]: Done :)

File: v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/package-info.java [hunk 21]
```diff
@@ -0,0 +1,21 @@
...
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+/**
+ * Package to hold template code.
+ */
+
+package com.google.cloud.teleport.v2.templates;
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/kafka-to-bigquery/src/main/java/com/google/cloud/teleport/v2/templates/package-info.java]: Done :)

**Comment by sabhyankar** [comment]
@chrlarsen  - Can you confirm all the changes were done? I see a single commit so it seems you squashed the commits already! (Thanks!)

File: v2/common/pom.xml [hunk 22]
```diff
@@ -0,0 +1,58 @@
...
+                </exclusion>
+            </exclusions>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.avro</groupId>
+            <artifactId>avro</artifactId>
+            <version>${avro.version}</version>
+        </dependency>
+        <dependency>
+            <groupId>com.google.cloud.teleport.v2</groupId>
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/common/pom.xml]: Not needed, removed

**Comment by chrlarsen** [comment]
@sabhyankar Yes all changes have been made and readmes for Kafka-to-BigQuery and BigQuery-to-GCS have also been added.

File: v2/common/src/test/java/com/google/cloud/teleport/v2/coders/FailsafeElementCoderTest.java [hunk 23]
```diff
@@ -0,0 +1,48 @@
...
+  private FailsafeElement<String, String> failsafeElement = FailsafeElement.of(originalPayload, originalPayload);
+
+  @Test
+  public void testEncoding() throws CoderException {
+    OutputStream out;
+    FailsafeElementCoder failsafeElementCoder = FailsafeElementCoder.of(STRING_CODER, STRING_CODER);
+    failsafeElementCoder.encode(null, out);
+
+  }
+
```
[the remark this answers is not in the document]
Comment (chrlarsen) [reply v2/common/src/test/java/com/google/cloud/teleport/v2/coders/FailsafeElementCoderTest.java]: Done, removed

**Comment by sabhyankar** [comment]
LGTM

**Pull request merged**